api/rtx/src/uvisor_malloc.c

```diff
@@ -0,0 +1,167 @@
+/* #define DPRINTF(fmt, ...) printf(fmt, ## __VA_ARGS__) */
+
+/* Declare this variable here, so uVisor initializes enough space for the
+ * mutexes, which are used in this file! */
+const uint32_t __uvisor_box_index_size = sizeof(RtxBoxIndex);
+
+extern RtxBoxIndex *const __uvisor_ps;
+extern void uvisor_malloc_init(void);
+
+static int is_kernel_initialized()
+{
+    static uint8_t kernel_running = 0;
+    return (kernel_running || (kernel_running = osKernelRunning()));
+}
+
+static int init_allocator()
+{
+    int ret = 0;
+    if (__uvisor_ps == NULL) {
+#if defined(UVISOR_PRESENT) && (UVISOR_PRESENT == 1)
+        return -1;
+#else
+        uvisor_malloc_init();
+#endif
+    }
+
+    if ((__uvisor_ps->mutex_id == NULL) && is_kernel_initialized()) {
+        /* point the mutex pointer to the data */
+        __uvisor_ps->mutex.mutex = &(__uvisor_ps->mutex_data);
+        /* create mutex if not already done */
+        __uvisor_ps->mutex_id = osMutexCreate(&(__uvisor_ps->mutex));
+        /* mutex failed to be created */
+        if (__uvisor_ps->mutex_id == NULL) {
+            return -1;
+        }
+    }
+
+    if (__uvisor_ps->index.active_heap == NULL) {
+        /* we need to initialize the process heap */
+        if (__uvisor_ps->index.box_heap != NULL) {
+            /* lock the mutex during initialization */
+            int kernel_initialized = is_kernel_initialized();
+            if (kernel_initialized) {
+                osMutexWait(__uvisor_ps->mutex_id, osWaitForever);
+            }
+            /* initialize the process heap */
+            UvisorAllocator allocator = uvisor_allocator_create_with_pool(
+                __uvisor_ps->index.box_heap,
+                __uvisor_ps->index.box_heap_size);
+            /* set the allocator */
+            ret = allocator ? 0 : -1;
+            __uvisor_ps->index.active_heap = allocator;
+            /* release the mutex */
+            if (kernel_initialized) {
+                osMutexRelease(__uvisor_ps->mutex_id);
+            }
+        }
+        else {
+            DPRINTF("uvisor_allocator: No process heap available!\n");
+            ret = -1;
+        }
+    }
+    return ret;
+}
+
+#define OP_MALLOC  0
+#define OP_REALLOC 1
+#define OP_FREE    2
+
+#define HEAP_ACTIVE  0
+#define HEAP_PROCESS 1
+
+static void * memory(void * ptr, size_t size, int heap, int operation)
+{
+    /* buffer the return value */
+    void * ret = NULL;
+    /* initialize allocator */
+    if (init_allocator()) {
+        return NULL;
+    }
+    /* check if we need to aquire the mutex */
+    int mutexed = (is_kernel_initialized() && (heap == HEAP_PROCESS));
+    void * allocator = (heap == HEAP_PROCESS) ?
+                       (__uvisor_ps->index.box_heap) :
+                       (__uvisor_ps->index.active_heap);
+
+    /* aquire the mutex if required */
+    if (mutexed) {
+        osMutexWait(__uvisor_ps->mutex_id, osWaitForever);
+    }
```

Comment:
Note: mutex use here is very coarse. Comment what exactly the mutex is protecting. 

We'd prefer finer grained locking, but since that's a big change, file an issue to make the locking finer grained. We'll revisit the issue later.